lib/util/regex.ts

```diff
@@ -19,21 +19,48 @@ try {
   RegEx = RegExp;
 }
 
+function extractFlags(flagSet: Set<string>, flags: string): Set<string> {
+  for (const flag of flags.split('').map((f) => f.toLowerCase())) {
+    flagSet.add(flag);
+  }
+  return flagSet;
+}
+
+function makeFlags(flagSet: Set<string>): string {
+  return [...flagSet].sort().join('');
+}
+
+function isStateful(flagSet: Set<string>): boolean {
+  return flagSet.has('g') || flagSet.has('y');
+}
+
 export function regEx(
   pattern: string | RegExp,
-  flags?: string | undefined,
+  withFlags?: string | undefined,
   useCache = true
 ): RegExp {
-  const key = flags ? `${pattern.toString()}:${flags}` : pattern.toString();
+  const flagSet = new Set<string>();
+  if (is.regExp(pattern)) {
+    extractFlags(flagSet, pattern.flags);
+  }
+  if (withFlags) {
+    extractFlags(flagSet, withFlags);
+  }
+  const flags = makeFlags(flagSet);
+  const isRegexStateful = isStateful(flagSet);
```

Comment:
we should add a throw statement here when useCache isn't false and regex is stateful. then run all tests and fix calling code.

so we don't need such complicated/ heavy code here

Reply:
If we throw, we'll have to explicitly set `useCache=false` for maybe half of the `regEx` calls, we can miss something in the test and get these exceptions in the production, and also I think it will be annoying DX when working with regexes in the future.

Instead, let's rewrite it to be less heavy maybe?

Reply:
What exactly is heavy from your perspective?

Reply:
> If we throw, we'll have to explicitly set `useCache=false` for maybe half of the `regEx` calls, we can miss something in the test and get these exceptions in the production, and also I think it will be annoying DX when working with regexes in the future.

We should be able to easily find all occurences via code search. As this can only occure on code. I think all user config regex have no stateful flags.
I just wanted to fix known places now and pass `useCache=false`.

> 
> Instead, let's rewrite it to be less heavy maybe?

skip flags computation when `useCache=false`

Reply:
Actually, I think `useCache` for regular expressions is the overkill. I think when we want to preserve compiled regex (non-stateful), we just move it to the module scope. When we need construct it from scratch every time, well, I think we can just create it couple more times.

All performance concerns I can remember don't include regular expressions as the main bottleneck. We have about `1024` calls of `regEx` function and I guess avarage regex is no more than 1Kb in memory, so seems like we're trying to win just 1 megabyte here in terms of memory.

Reply:
It's very easy to create memory leak with `useCache`, it requires just construct regex from string like `depName` or, even better, from `version` and then just forget to make `useCache=false`.

Wasn't the regex caching created as a premature optimization?

Reply:
it not the memory, it's the recreation of regex in nested loops thousands of times.

the cache improved the overall runtime of renovate, especially on repos with lots of files to scan by renovate.

Reply:
Okay, sorry for my hot takes then. I'm reworking this PR now.